From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: NONPLAYT <76615486+NONPLAYT@users.noreply.github.com>
Date: Sun, 6 Jul 2025 01:55:37 +0300
Subject: [PATCH] C2ME: Optimize world gen math

This patch is based on following mixins:
* "com/ishland/c2me/opts/math/mixin/MixinChunkPos.java"
* "com/ishland/c2me/opts/worldgen/vanilla/mixin/structure_weight_sampler/MixinStructureWeightSampler.java"
By: ishland <ishlandmc@yeah.net>
As part of: C2ME-fabric (https://github.com/RelativityMC/C2ME-fabric)
Licensed under: MIT

diff --git a/net/minecraft/world/level/ChunkPos.java b/net/minecraft/world/level/ChunkPos.java
index 4caaed6d36a6d12279d68e974a2e8381ba84a951..16b5dac678790cebc8fb577fa2f1c7a05aaeb85b 100644
--- a/net/minecraft/world/level/ChunkPos.java
+++ b/net/minecraft/world/level/ChunkPos.java
@@ -110,7 +110,12 @@ public class ChunkPos {
 
     @Override
     public boolean equals(Object other) {
-        return this == other || other instanceof ChunkPos chunkPos && this.x == chunkPos.x && this.z == chunkPos.z;
+        // DivineMC start - C2ME: Optimize world gen math
+        if (other == this) return true;
+        if (other == null || other.getClass() != this.getClass()) return false;
+        ChunkPos thatPos = (ChunkPos) other;
+        return this.x == thatPos.x && this.z == thatPos.z;
+        // DivineMC end - C2ME: Optimize world gen math
     }
 
     public int getMiddleBlockX() {
diff --git a/net/minecraft/world/level/levelgen/Beardifier.java b/net/minecraft/world/level/levelgen/Beardifier.java
index 5e1d88ef91e585a34e9213da923df796dd866e03..c9ddfc8670614c2d8629066b0cc805d18e4f662f 100644
--- a/net/minecraft/world/level/levelgen/Beardifier.java
+++ b/net/minecraft/world/level/levelgen/Beardifier.java
@@ -168,8 +168,14 @@ public class Beardifier implements DensityFunctions.BeardifierOrMarker {
     }
 
     private static double getBuryContribution(double x, double y, double z) {
-        double len = Mth.length(x, y, z);
-        return Mth.clampedMap(len, 0.0, 6.0, 1.0, 0.0);
+        // DivineMC start - C2ME: Optimize world gen math
+        double len = Math.sqrt(x * x + y * y + z * z);
+        if (len > 6.0) {
+            return 0.0;
+        } else {
+            return 1.0 - len / 6.0;
+        }
+        // DivineMC end - C2ME: Optimize world gen math
     }
 
     private static double getBeardContribution(int x, int y, int z, int height) {
diff --git a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
index 199a9c4fba5fc80f04524a188270fdc9014f7950..952d00151cc6c295f37bcc499ddcb8155185c79c 100644
--- a/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
+++ b/net/minecraft/world/level/levelgen/NoiseBasedChunkGenerator.java
@@ -73,7 +73,10 @@ public final class NoiseBasedChunkGenerator extends ChunkGenerator {
             if (SharedConstants.DEBUG_DISABLE_FLUID_GENERATION) {
                 return fluidStatus2;
             } else {
-                return y < Math.min(-54, seaLevel) ? fluidStatus : fluidStatus1;
+                // DivineMC start - C2ME: Optimize world gen math
+                final int min = Math.min(-54, seaLevel);
+                return y < min ? fluidStatus : fluidStatus1;
+                // DivineMC end - C2ME: Optimize world gen math
             }
         };
     }
